Implement spec-compliant chunk-stream-id serialization/deserialization

dmorn opened the pull request:
The RTMP server was crashing when receiving streams from certain hardware encoders with the following error:

```
[warning] Unexpected header type when combining body chunks: 1
[error] GenServer #PID<0.1014.0> terminating
** (KeyError) key :timestamp not found in: nil
    (membrane_rtmp_plugin 0.29.1) lib/membrane_rtmp_plugin/rtmp/header.ex:114: Membrane.RTMP.Header.deserialize/2
    (membrane_rtmp_plugin 0.29.1) lib/membrane_rtmp_plugin/rtmp/message_parser.ex:180: Membrane.RTMP.MessageParser.read_frame/3
```
The crash was caused by **incomplete support for RTMP chunk interleaving**. Further, only chunk-stream-id singe byte serialization/deserialization was implemented.


This PR adds support for 1/2/3 byte type stream chunk identifiers and handles the situation where chunks from different streams are interleaved together. Fully backward compatible.

dmorn committed:
Implement spec-compliant chunk-stram-id serde

philipgiuliani committed:
Fix credo warnings

dmorn committed:
Handle chunked stream ids also for chunked messages

dmorn committed:
Checkpoint

dmorn committed:
Fix infinite recursion

dmorn committed:
Revert "Fix infinite recursion"

This reverts commit dde4b1279fd4530229daf95ac90b12e754656cb5.

dmorn committed:
Revert "Checkpoint"

This reverts commit 16e9fc9f641544142afbe56a42d1030f01fc7115.

dmorn committed:
Revert "Handle chunked stream ids also for chunked messages"

This reverts commit faaa13cf6be1002cd34e5fc820ad88d0ab287026.

dmorn committed:
Fix chunked message interleaving

dmorn committed:
Protect against duplicate connections

dmorn committed:
Format

dmorn committed:
format

dmorn committed:
Handle multiple chunks in the same packet (fix ffmpeg)

dmorn committed:
Merge branch 'master' into chunk-stream-id

# Conflicts:
#	lib/membrane_rtmp_plugin/rtmp/header.ex

dmorn committed:
Fix credo nesting issue in continue_partial_message

dmorn committed:
Fix struct update warnings in Header.deserialize_message_header/3

Pattern match previous_header as %__MODULE__{} in the :type_3 clause
to satisfy the type checker for struct updates.

varsill commented on lib/membrane_rtmp_plugin/rtmp/source/source.ex:
```diff
@@ -176,18 +176,31 @@ defmodule Membrane.RTMP.Source do
+      )
+      when app == state.app and stream_key == state.stream_key do
+    # Duplicate connection on same app/stream_key - ignore it
+    Logger.warning("Duplicate client connection detected on /#{app}/#{stream_key}, ignoring")
```
Perhaps we could log the `client_ref` here so that it would be possible to distinguish which one gets discarded?

varsill reviewed (approved):
Thanks for the PR!
I left one minor comment.
I also wonder if it would be possible for you to add a single integration test checking the interleaved chunks behaviour (especially if you already have some stream dump we could use as a fixture).

dmorn committed:
Log duplicate RTMP client refs

dmorn committed:
Add 4K Blackmagic RTMP parser regression test

dmorn commented:
Thanks! I addressed the logging comment by including both the discarded `client_ref` and the already-handled one in the warning.

I also added an integration-style parser regression test based on a real Blackmagic Web Presenter 4K capture. The fixture stores post-handshake client→server RTMP TCP reads with original read boundaries preserved, which turned out to be important for reproducing the issue. I kept it constrained (~800 KB) instead of committing a full 4K media recording.

The interesting bit is that replaying this fixture leaves leftover buffered data on `master`, while this branch fully consumes it, so it gives us a real-world regression check for the parser behavior we’re fixing.

dmorn committed:
Fix Header.deserialize Dialyzer spec

dmorn committed:
Format Header.deserialize spec

varsill commented:
Thanks for contribution @dmorn ! I will release it together with https://github.com/membraneframework/membrane_rtmp_plugin/pull/122, is that OK?

dmorn commented:
Of course! I'll prepare in next week (tuesday) @varsill!
